coord/src/rmr/rmr.c

```diff
@@ -710,20 +707,35 @@ void MRIterator_WaitDone(MRIterator *it, bool mayBeIdle) {
     while (MRIteratorCallback_GetNumInProcess(it)) {
       usleep(1000);
     }
-  } else {
-    // Wait until all the commands are done (it->ctx.pending == 0)
-    MRChannel_WaitClose(it->ctx.chan);
+    // If we have no pending shards, we are done.
+    if (!it->ctx.pending) return;
+    // If we have pending (not depleted) shards, trigger `FT.CURSOR DEL` on them
+    it->ctx.inProcess = it->ctx.pending;
+    // Change the root command to DEL for each pending shard
+    for (size_t i = 0; i < it->len; i++) {
+      MRCommand *cmd = &it->cbxs[i].cmd;
+      if (!cmd->depleted) {
+        // assert(!strcmp(cmd->strs[1], "READ"));
+        cmd->rootCommand = C_DEL;
+        strcpy(cmd->strs[1], "DEL");
+        cmd->lens[1] = 3;
+      }
+    }
+    // Send the DEL commands, and wait for them to be done
+    RQ_Push(rq_g, iterManualNextCb, it);
   }
+  // Wait until all the commands are done (it->ctx.pending == 0)
+  MRChannel_WaitClose(it->ctx.chan);
 }
 
+// Assumes no other thread is using the iterator, the channel, or any of the commands and contexts
 void MRIterator_Free(MRIterator *it) {
-  if (!it) return;
   for (size_t i = 0; i < it->len; i++) {
     MRCommand_Free(&it->cbxs[i].cmd);
   }
   MRReply *reply;
-  while((reply = MRChannel_ForcePop(it->ctx.chan))){
-      MRReply_Free(reply);
+  while ((reply = MRChannel_UnsafeForcePop(it->ctx.chan))) {
```

Comment:
How do you know we can call it without the lock? Is it because the free is only performed after the `MRIterator_WaitDone` is finished? If so maybe we can merge them to a single function?

Reply:
Yes that’s why. Seems wasteful to lock and unlock for every link in the list when there is no race.
I don’t think we should merge them. First the free function is used in another place when we don’t need to wait and even shouldn’t. Second we probably going to need some similar function for freeing the iterators without waiting (with some reference managing)

Reply:
But if the free function is called somewhere else without wait, does it still safe to call it without the lock?

Reply:
Currently, it is called once after the wait from the `rpnet` destructor and once if there is some error while creating it. In both cases it is safe, and in the future, I believe we'll want to use it as the free callback from some ref-management system.
In the `rpnet` destructor we also do some additional work after waiting that is unrelated to freeing (collecting profiling results from the channel after we waited). So we probably have to wait before the destruction anyway.